File: internal/manager/trustedrouter/channel.go
package trustedrouter

import (
	"context"
	"encoding/json"
	"errors"
	"io"
	"net"
	"net/http"
	"sync"
	"time"

	"github.com/codeasier/mtls-router/internal/manager/apikeyusage"
	"github.com/codeasier/mtls-router/internal/manager/discovery"
	"github.com/codeasier/mtls-router/internal/manager/modelcatalog"
	"github.com/codeasier/mtls-router/internal/manager/process"
	"github.com/codeasier/mtls-router/internal/manager/protocol"
	"github.com/codeasier/mtls-router/internal/manager/state"
)

const versionBodyLimit = 64 << 10

// Channel validates router identity and fetches a catalog over one TCP
// connection. A server-requested redial is rejected before a second request can
// transmit its Authorization header.
type Channel struct {
	Simplify        bool
	DialContext     func(context.Context, string, string) (net.Conn, error)
	ValidateProcess func(process.Identity, string) (process.Status, error)
}

// Fetch performs the channel-bound /version and /v1/models exchange.
func (c Channel) Fetch(ctx context.Context, listener Listener, trusted discovery.Result, apiKey string) ([]string, *protocol.Error) {
	transport, bindErr := c.bind(ctx, listener, trusted)
	if bindErr != nil {
		return nil, bindErr
	}
	defer transport.CloseIdleConnections()
	models, fetchErr := modelcatalog.New(transport, c.Simplify).Fetch(ctx, modelcatalog.Request{
		URL: listener.APIBaseURL + "/models", APIKey: apiKey,
	})
	if fetchErr != nil {
		return nil, &protocol.Error{Code: modelcatalog.CodeOf(fetchErr), Message: fetchErr.Error()}
	}
	return models, nil
}

// FetchUsage performs the channel-bound /version and /v1/usage exchange.
func (c Channel) FetchUsage(ctx context.Context, listener Listener, trusted discovery.Result, period apikeyusage.Period, apiKey string) (apikeyusage.Snapshot, *protocol.Error) {
	transport, bindErr := c.bind(ctx, listener, trusted)
	if bindErr != nil {
		return apikeyusage.Snapshot{}, bindErr
	}
	defer transport.CloseIdleConnections()
	snapshot, fetchErr := apikeyusage.New(transport).Fetch(ctx, apikeyusage.Request{
		URL: listener.APIBaseURL + "/usage", Period: period, APIKey: apiKey,
	})
	if fetchErr != nil {
		return apikeyusage.Snapshot{}, &protocol.Error{Code: apikeyusage.CodeOf(fetchErr), Message: fetchErr.Error()}
	}
	return snapshot, nil
}

func (c Channel) bind(ctx context.Context, listener Listener, trusted discovery.Result) (*http.Transport, *protocol.Error) {
	dial := c.DialContext
	if dial == nil {
		dial = (&net.Dialer{Timeout: 5 * time.Second}).DialContext
	}
	validate := c.ValidateProcess
	if validate == nil {
		validate = process.Validate
	}

	var dialMu sync.Mutex
	dialed := false
	transport := &http.Transport{
		Proxy: nil,
		DialContext: func(dialCtx context.Context, network, address string) (net.Conn, error) {
			dialMu.Lock()
			if dialed || network != "tcp" || address != listener.Authority {
				dialMu.Unlock()
				return nil, errors.New("trusted router connection cannot redial")
			}
			dialed = true
			dialMu.Unlock()
			return dial(dialCtx, network, address)
		},
		MaxConnsPerHost:     1,
		MaxIdleConns:        1,
		MaxIdleConnsPerHost: 1,
		IdleConnTimeout:     5 * time.Second,
		DisableCompression:  true,
	}
	httpClient := &http.Client{
		Transport: transport,
		Timeout:   15 * time.Second,
		CheckRedirect: func(*http.Request, []*http.Request) error {
			return http.ErrUseLastResponse
		},
	}

	versionRequest, err := http.NewRequestWithContext(ctx, http.MethodGet, listener.RouterBaseURL+"/version", nil)
	if err != nil {
		transport.CloseIdleConnections()
		return nil, discoveryFailed()
	}
	response, err := httpClient.Do(versionRequest)
	if err != nil {
		transport.CloseIdleConnections()
		return nil, discoveryFailed()
	}
	if response.StatusCode != http.StatusOK {
		response.Body.Close()
		transport.CloseIdleConnections()
		return nil, discoveryFailed()
	}
	body, err := io.ReadAll(io.LimitReader(response.Body, versionBodyLimit+1))
	closeErr := response.Body.Close()
	if err != nil || closeErr != nil || len(body) > versionBodyLimit {
		transport.CloseIdleConnections()
		return nil, discoveryFailed()
	}
	var versionInfo discovery.Version
	decoderErr := json.Unmarshal(body, &versionInfo)
	if decoderErr != nil || !versionMatches(versionInfo, trusted) {
		transport.CloseIdleConnections()
		return nil, staleCatalog()
	}
	identity := process.Identity{PID: trusted.State.PID, StartedAt: trusted.State.ProcessStartedAt, Executable: trusted.State.ProcessExecutable}
	status, validateErr := validate(identity, trusted.State.BinaryPath)
	if validateErr != nil || status != process.StatusGenuine {
		transport.CloseIdleConnections()
		return nil, staleCatalog()
	}
	return transport, nil
}

func versionMatches(remote discovery.Version, trusted discovery.Result) bool {
	value := trusted.State
	return remote.PID > 0 && remote.PID == value.PID &&
		remote.DeploymentID == value.DeploymentID && remote.DeploymentID == trusted.Version.DeploymentID &&
		remote.ManagementProtocolVersion == value.ManagementProtocolVersion &&
		remote.ManagementProtocolVersion == trusted.Version.ManagementProtocolVersion
}

func trustedStateMatches(found discovery.Result, listener Listener, deploymentID, protocolVersion string) bool {
	if found.Classification != discovery.DesktopOwned && found.Classification != discovery.ExternalCompatible && found.Classification != discovery.Degraded {
		return false
	}
	value := found.State
	return (value.Owner == "cli" || value.Owner == "desktop") && value.PID > 0 &&
		value.ListenAddr == listener.RouterBaseURL && found.ListenAddr == listener.RouterBaseURL &&
		value.DeploymentID == deploymentID && found.Version.DeploymentID == deploymentID &&
		value.ManagementProtocolVersion == protocolVersion && found.Version.ManagementProtocolVersion == protocolVersion
}

func stateFromStart(value state.RouterState, listener Listener, deploymentID, protocolVersion string) bool {
	return value.PID > 0 && (value.Owner == "cli" || value.Owner == "desktop") &&
		value.ListenAddr == listener.RouterBaseURL && value.DeploymentID == deploymentID &&
		value.ManagementProtocolVersion == protocolVersion
}

func discoveryFailed() *protocol.Error {
	return &protocol.Error{Code: protocol.CodeModelDiscoveryFailed, Message: "model catalog request failed"}
}

func staleCatalog() *protocol.Error {
	return &protocol.Error{Code: protocol.CodeModelCatalogStale, Message: "trusted router identity changed"}
}

File: desktop/src-tauri/src/manager_core/trustedrouter/channel.rs
use std::io;
use std::net::TcpStream;
use std::sync::Arc;
use std::time::{Duration, Instant};

use serde::Deserialize;

use crate::manager_core::apikeyusage::{self, Period, Snapshot};
use crate::manager_core::errors::closed_error;
use crate::manager_core::httpget::{self, HttpTransport};
use crate::manager_core::modelcatalog;
use crate::manager_core::process::{self, Identity, ProcessError, Status};
use crate::protocol::{ErrorCode, ProtocolError};

use super::listener::Listener;
use super::{Discovery, RemoteVersion};

pub const VERSION_BUDGET: Duration = Duration::from_secs(15);
const VERSION_BODY_LIMIT: usize = 64 << 10;
const DIAL_TIMEOUT: Duration = Duration::from_secs(5);

type DialFn = Arc<dyn Fn(&str) -> io::Result<TcpStream> + Send + Sync>;
type ValidateFn = Arc<dyn Fn(&Identity, &str) -> Result<Status, ProcessError> + Send + Sync>;

#[derive(Clone)]
pub struct Channel {
    pub simplify: bool,
    pub dial: Option<DialFn>,
    pub validate_process: Option<ValidateFn>,
}

impl Default for Channel {
    fn default() -> Self {
        Self {
            simplify: false,
            dial: None,
            validate_process: None,
        }
    }
}

impl Channel {
    pub fn fetch(
        &self,
        listener: &Listener,
        trusted: &Discovery,
        api_key: &str,
        deadline: Instant,
    ) -> Result<Vec<String>, ProtocolError> {
        let mut transport = self.bind(listener, trusted, deadline)?;
        let url = format!("{}/models", listener.api_base_url);
        modelcatalog::Client::new(self.simplify)
            .fetch_with(
                &mut transport,
                deadline,
                modelcatalog::Request {
                    url,
                    api_key: api_key.to_owned(),
                },
            )
            .map_err(modelcatalog::CatalogError::into_protocol)
    }

    pub fn fetch_usage(
        &self,
        listener: &Listener,
        trusted: &Discovery,
        period: Period,
        api_key: &str,
        deadline: Instant,
    ) -> Result<Snapshot, ProtocolError> {
        let bind_deadline = min_deadline(deadline, VERSION_BUDGET);
        let mut transport = self.bind(listener, trusted, bind_deadline)?;
        let usage_deadline = min_deadline(deadline, apikeyusage::REQUEST_TIMEOUT);
        let url = format!("{}/usage", listener.api_base_url);
        apikeyusage::Client::new()
            .fetch_with(
                &mut transport,
                usage_deadline,
                apikeyusage::Request {
                    url,
                    period,
                    api_key: api_key.to_owned(),
                },
            )
            .map_err(apikeyusage::UsageError::into_protocol)
    }

    fn bind(
        &self,
        listener: &Listener,
        trusted: &Discovery,
        deadline: Instant,
    ) -> Result<BoundTransport, ProtocolError> {
        let mut transport = BoundTransport {
            listener: listener.clone(),
            stream: None,
            dialed: false,
            closed: false,
            dial: self.dial.clone(),
        };
        let response = transport
            .get("/version", None, deadline, VERSION_BODY_LIMIT + 1)
            .map_err(|_| discovery_failed())?;
        if response.status != 200 {
            return Err(discovery_failed());
        }
        if response.body.len() > VERSION_BODY_LIMIT {
            return Err(discovery_failed());
        }
        let remote: WireVersion =
            serde_json::from_slice(&response.body).map_err(|_| stale_catalog())?;
        if !version_matches(
            &RemoteVersion {
                pid: remote.pid,
                deployment_id: remote.deployment_id,
                management_protocol_version: remote.management_protocol_version,
            },
            trusted,
        ) {
            return Err(stale_catalog());
        }
        let identity = Identity {
            pid: trusted.state.pid,
            started_at: trusted.state.process_started_at.clone(),
            executable: trusted.state.process_executable.clone(),
        };
        let status = match &self.validate_process {
            Some(validate) => validate(&identity, &trusted.state.binary_path),
            None => process::validate(&identity, &trusted.state.binary_path),
        };
        if !matches!(status, Ok(Status::Genuine)) {
            return Err(stale_catalog());
        }
        Ok(transport)
    }
}

#[derive(Deserialize)]
struct WireVersion {
    #[serde(default)]
    pid: i32,
    #[serde(default)]
    deployment_id: String,
    #[serde(default)]
    management_protocol_version: String,
}

pub(super) struct BoundTransport {
    listener: Listener,
    stream: Option<TcpStream>,
    dialed: bool,
    closed: bool,
    dial: Option<DialFn>,
}

impl HttpTransport for BoundTransport {
    fn get(
        &mut self,
        path: &str,
        authorization: Option<&str>,
        deadline: Instant,
        max_body: usize,
    ) -> Result<httpget::HttpResponse, ()> {
        if self.closed || (self.dialed && self.stream.is_none()) {
            return Err(());
        }
        if self.stream.is_none() {
            if self.dialed {
                return Err(());
            }
            self.dialed = true;
            let stream = match &self.dial {
                Some(dial) => dial(&self.listener.authority).map_err(|_| ())?,
                None => {
                    let dial_deadline = min_deadline(deadline, DIAL_TIMEOUT);
                    httpget::connect(&self.listener.authority, dial_deadline)?
                }
            };
            self.stream = Some(stream);
        }
        let stream = self.stream.as_mut().ok_or(())?;
        let response = httpget::request(
            stream,
            &self.listener.authority,
            path,
            authorization,
            deadline,
            max_body,
        )?;
        if response.connection_close {
            self.closed = true;
            self.stream = None;
        }
        Ok(response)
    }
}

pub fn version_matches(remote: &RemoteVersion, trusted: &Discovery) -> bool {
    remote.pid > 0
        && remote.pid == trusted.state.pid
        && remote.deployment_id == trusted.state.deployment_id
        && remote.deployment_id == trusted.version.deployment_id
        && remote.management_protocol_version == trusted.state.management_protocol_version
        && remote.management_protocol_version == trusted.version.management_protocol_version
}

pub fn trusted_state_matches(
    found: &Discovery,
    listener: &Listener,
    deployment_id: &str,
    protocol_version: &str,
) -> bool {
    if !found.classification.is_trusted() {
        return false;
    }
    let value = &found.state;
    (value.owner == "cli" || value.owner == "desktop")
        && value.pid > 0
        && value.listen_addr == listener.router_base_url
        && found.listen_addr == listener.router_base_url
        && value.deployment_id == deployment_id
        && found.version.deployment_id == deployment_id
        && value.management_protocol_version == protocol_version
        && found.version.management_protocol_version == protocol_version
}

pub fn state_from_start(
    value: &super::StartedIdentity,
    listener: &Listener,
    deployment_id: &str,
    protocol_version: &str,
) -> bool {
    value.pid > 0
        && (value.owner == "cli" || value.owner == "desktop")
        && value.listen_addr == listener.router_base_url
        && value.deployment_id == deployment_id
        && value.management_protocol_version == protocol_version
}

pub fn discovery_failed() -> ProtocolError {
    closed_error(
        ErrorCode::ModelDiscoveryFailed,
        "model catalog request failed",
    )
}

pub fn stale_catalog() -> ProtocolError {
    closed_error(
        ErrorCode::ModelCatalogStale,
        "trusted router identity changed",
    )
}

pub fn min_deadline(parent: Instant, budget: Duration) -> Instant {
    let local = Instant::now() + budget;
    if parent < local {
        parent
    } else {
        local
    }
}
